File: app/src/main/java/com/slapshotapps/dragonshockey/activities/admin/EditStatsActivity.java
package com.slapshotapps.dragonshockey.activities.admin;

import android.os.Bundle;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.appcompat.widget.Toolbar;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.slapshotapps.dragonshockey.Config;
import com.slapshotapps.dragonshockey.R;
import com.slapshotapps.dragonshockey.Utils.DragonsHockeyIntents;
import com.slapshotapps.dragonshockey.Utils.RosterUtils;
import com.slapshotapps.dragonshockey.activities.admin.adapter.AdminEditsStatsAdapter;
import com.slapshotapps.dragonshockey.activities.admin.viewmodels.AdminStatsViewModel;
import com.slapshotapps.dragonshockey.models.GameStats;
import com.slapshotapps.dragonshockey.models.Player;
import com.slapshotapps.dragonshockey.models.PlayerGameStats;
import com.slapshotapps.dragonshockey.observables.AdminObserver;
import java.util.ArrayList;
import java.util.List;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

public class EditStatsActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private Subscription subscription;
    private FirebaseDatabase firebaseDatabase;
    private PlayerGameStats playerGameStats;
    private AdminEditsStatsAdapter adminEditsStatsAdapter;
    private int gameID;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_edit_stats_auth);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        recyclerView = findViewById(R.id.admin_stats_recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        firebaseDatabase = FirebaseDatabase.getInstance();

        gameID = getIntent().getIntExtra(DragonsHockeyIntents.EXTRA_GAME_ID, 0);
    }

    @Override
    protected void onResume() {
        super.onResume();

        if (adminEditsStatsAdapter == null) {
            subscription = AdminObserver.getPlayerStatsForGame(firebaseDatabase, gameID)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(stats -> {
                    playerGameStats = stats;

                    adminEditsStatsAdapter = new AdminEditsStatsAdapter(getViewModel());
                    recyclerView.setAdapter(adminEditsStatsAdapter);
                }, throwable -> new AlertDialog.Builder(EditStatsActivity.this).setMessage(
                    "Unable to retrieve stats information")
                    .setPositiveButton("Ok", (dialogInterface, i) -> finish())
                    .show());
        }
    }

    @Override
    protected void onPause() {
        super.onPause();

        if (subscription != null) {
            subscription.unsubscribe();
            subscription = null;
        }
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();

        if (adminEditsStatsAdapter.statsChanged()) {

            ArrayList<AdminStatsViewModel> viewModel = adminEditsStatsAdapter.getStats();

            if (playerGameStats.isKeyValid()) {
                firebaseDatabase.getReference()
                    .child(Config.GAME_STATS)
                    .child(playerGameStats.playerStatsKey)
                    .setValue(getGameStats(viewModel));
            } else {
                DatabaseReference newGameResultRef =
                    firebaseDatabase.getReference().child(Config.GAME_STATS).push();
                newGameResultRef.setValue(getGameStats(viewModel));
            }
        }
    }

    private ArrayList<AdminStatsViewModel> getViewModel() {
        ArrayList<AdminStatsViewModel> statsViewModel = new ArrayList<AdminStatsViewModel>();

        for (Player player : playerGameStats.players) {

            GameStats.Stats statsForPlayer =
                playerGameStats.playerGameStats.getPlayerStats(player.playerID);

            if (statsForPlayer == null) {
                statsForPlayer = new GameStats.Stats();
            }

            AdminStatsViewModel viewModel =
                new AdminStatsViewModel.AdminStatsVMBuilder().playerName(
                    RosterUtils.getFullName(player))
                    .playerID(player.playerID)
                    .playerNumber(player.number)
                    .goals(statsForPlayer.getGoals())
                    .assists(statsForPlayer.getAssists())
                    .present(statsForPlayer.getPresent())
                    .penaltyMinutes(statsForPlayer.getPenaltyMinutes())
                    .position(player.getPosition())
                    .goalsAgainst(statsForPlayer.getGoalsAgainst())
                    .build();

            statsViewModel.add(viewModel);
        }

        return statsViewModel;
    }

    private GameStats getGameStats(List<AdminStatsViewModel> adminStatsViewModelList) {
        GameStats gameStats = new GameStats();

        gameStats.setGameID(gameID);
        gameStats.setGameStats(new ArrayList<>());

        for (AdminStatsViewModel adminStatsViewModel : adminStatsViewModelList) {
            GameStats.Stats stats = new GameStats.Stats();
            stats.setPlayerID(adminStatsViewModel.getPlayerID());
            stats.setAssists(Integer.valueOf(adminStatsViewModel.getAssists()));
            stats.setGoals(Integer.valueOf(adminStatsViewModel.getGoals()));
            stats.setPresent(adminStatsViewModel.getPresence());
            stats.setPenaltyMinutes(Integer.valueOf(adminStatsViewModel.getPenaltyMinutes()));
            stats.setGoalsAgainst(Integer.valueOf(adminStatsViewModel.getGoalsAgainst()));

            gameStats.getGameStats().add(stats);
        }

        return gameStats;
    }
}

File: app/src/main/java/com/slapshotapps/dragonshockey/activities/admin/EditStatsActivity.kt
package com.slapshotapps.dragonshockey.activities.admin

import android.content.DialogInterface
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.database.FirebaseDatabase
import com.slapshotapps.dragonshockey.Config
import com.slapshotapps.dragonshockey.R
import com.slapshotapps.dragonshockey.activities.admin.adapter.AdminEditsStatsAdapter
import com.slapshotapps.dragonshockey.activities.admin.viewmodels.AdminStatsViewModel
import com.slapshotapps.dragonshockey.activities.admin.viewmodels.AdminStatsViewModel.AdminStatsVMBuilder
import com.slapshotapps.dragonshockey.databinding.ActivityEditStatsAuthBinding
import com.slapshotapps.dragonshockey.models.GameStats
import com.slapshotapps.dragonshockey.models.PlayerGameStats
import com.slapshotapps.dragonshockey.observables.AdminObserver
import com.slapshotapps.dragonshockey.utils.DragonsHockeyIntents
import com.slapshotapps.dragonshockey.utils.RosterUtils
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

class EditStatsActivity : AppCompatActivity() {
    private var binding: ActivityEditStatsAuthBinding? = null
    private var subscription: Subscription? = null
    private var firebaseDatabase: FirebaseDatabase? = null
    private var playerGameStats: PlayerGameStats? = null
    private var adminEditsStatsAdapter: AdminEditsStatsAdapter? = null
    private var gameID = 0
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val localBinding = ActivityEditStatsAuthBinding.inflate(layoutInflater)
        binding = localBinding
        setContentView(localBinding.root)
        setSupportActionBar(localBinding.toolbar)

        localBinding.editStatsContainer.adminStatsRecyclerView.layoutManager = LinearLayoutManager(this)
        firebaseDatabase = FirebaseDatabase.getInstance()
        gameID = intent.getIntExtra(DragonsHockeyIntents.EXTRA_GAME_ID, 0)
    }

    override fun onResume() {
        super.onResume()
        if (adminEditsStatsAdapter == null) {
            subscription = AdminObserver.getPlayerStatsForGame(firebaseDatabase, gameID)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({ stats: PlayerGameStats? ->
                        playerGameStats = stats
                        adminEditsStatsAdapter = AdminEditsStatsAdapter(viewModel)
                        binding?.editStatsContainer?.adminStatsRecyclerView?.adapter = adminEditsStatsAdapter
                    }) { _: Throwable? ->
                        AlertDialog.Builder(this@EditStatsActivity).setMessage(
                                "Unable to retrieve stats information")
                                .setPositiveButton("Ok") { _: DialogInterface?, _: Int -> finish() }
                                .show()
                    }
        }
    }

    override fun onPause() {
        super.onPause()
        if (subscription != null) {
            subscription!!.unsubscribe()
            subscription = null
        }
    }

    override fun onBackPressed() {
        super.onBackPressed()
        if (adminEditsStatsAdapter!!.statsChanged()) {
            val viewModel = adminEditsStatsAdapter!!.getPlayerStats()
            if (playerGameStats!!.isKeyValid) {
                firebaseDatabase!!.reference
                        .child(Config.GAME_STATS)
                        .child(playerGameStats!!.playerStatsKey!!)
                        .setValue(getGameStats(viewModel))
            } else {
                val newGameResultRef = firebaseDatabase!!.reference.child(Config.GAME_STATS).push()
                newGameResultRef.setValue(getGameStats(viewModel))
            }
        }
    }

    private val viewModel: ArrayList<AdminStatsViewModel>
        get() {
            val statsViewModel = ArrayList<AdminStatsViewModel>()
            playerGameStats?.let {
                it.players?.forEach { player ->
                    it.playerGameStats?.getPlayerStats(player.playerID)?.let { statsForPlayer ->
                        val viewModel = AdminStatsVMBuilder().playerName(
                                RosterUtils.getFullName(player))
                                .playerID(player.playerID)
                                .playerNumber(player.number)
                                .goals(statsForPlayer.goals)
                                .assists(statsForPlayer.assists)
                                .present(statsForPlayer.present)
                                .penaltyMinutes(statsForPlayer.penaltyMinutes)
                                .position(player.getPlayerPosition())
                                .goalsAgainst(statsForPlayer.goalsAgainst)
                                .build()
                        statsViewModel.add(viewModel)
                    }
                }
            }

            return statsViewModel
        }

    private fun getGameStats(adminStatsViewModelList: List<AdminStatsViewModel>): GameStats {
        val gameStats = GameStats()
        gameStats.gameID = gameID
        gameStats.gameStats = ArrayList()
        for (adminStatsViewModel in adminStatsViewModelList) {
            val stats = GameStats.Stats()
            stats.playerID = adminStatsViewModel.playerID
            stats.assists = Integer.valueOf(adminStatsViewModel.assists)
            stats.goals = Integer.valueOf(adminStatsViewModel.goals)
            stats.present = adminStatsViewModel.presence
            stats.penaltyMinutes = Integer.valueOf(adminStatsViewModel.penaltyMinutes)
            stats.goalsAgainst = Integer.valueOf(adminStatsViewModel.goalsAgainst)
            gameStats.gameStats?.add(stats)
        }
        return gameStats
    }
}